### PR TITLE
SALTO-858 - Convert referenceTo to reference expressions

### PR DESCRIPTION
**Pull request by netama**
Note: This needs to be run on a clean environment in order for existing annotations to be converted to references.

### CODE REVIEW

File: packages/salesforce-adapter/src/filters/reference_annotations.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+export const apiNameToElemID = (elements: Element[]): Record<string, ElemID> => (
+  _(elements)
+    .filter(isObjectType)
+    .map(e => [apiName(e), e.elemID])
+    .fromPairs()
+    .value()
+)
+
+/**
+ * Convert referenceTo and foreignKeyDomain annotations into reference expressions.
```
Comment (adamro) [inline packages/salesforce-adapter/src/filters/reference_annotations.ts:68]: you currently only did it for referenceTo, right?

File: packages/salesforce-adapter/src/filters/reference_annotations.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    }
+    return ref
+  }
+
+  elements
+    .filter(isObjectType)
+    .forEach((obj: ObjectType) => (Object.values(obj.fields)
+      .filter((field: Field) => field.annotations[annotationName] !== undefined)
+      .forEach((field: Field): void => {
+        field.annotations[annotationName] = field.annotations[annotationName]
```
Comment (adamro) [inline packages/salesforce-adapter/src/filters/reference_annotations.ts:53]: It looks like there's an assumption `field.annotations[annotationName]` is a string[] or ReferenceExpression[] at this stage. 
If it's not an array `.map` will fail. If it's an array but not of those types then the map will do nothing.
Probably need to add a `_.isArray` check (and maybe also handle the non-array case).

File: packages/salesforce-adapter/src/filters/reference_annotations.ts (hunk 1 above)
Comment (netama) [reply packages/salesforce-adapter/src/filters/reference_annotations.ts:68]: I'm going to push a change with `foreignKeyDomain`, this was a placeholder because there was a competing PR that introduced the `foreignKeyDomain` logic

**Comment by netama** [comment]
yes, the filter order was temporary to avoid interrupting your work :) I was going to check with you when would be a good time to move it before the other filters and changing them to rely on it.

File: packages/salesforce-adapter/src/filters/reference_annotations.ts (hunk 2 above)
Comment (netama) [reply packages/salesforce-adapter/src/filters/reference_annotations.ts:53]: you're right, I forgot to add `makeArray`, thanks!

File: packages/salesforce-adapter/src/transformers/transformer.ts [hunk 3]
```diff
@@ -1312,5 +1312,9 @@ export const getLookUpName: GetLookupNameFunc = ({ ref, path, field }) => {
     const isRelative = lookUpRelative(field, path)
     return apiName(ref.value, isRelative)
   }
+  // TODO remove - demonstrating the e2e issue
```
Comment (netama) [inline packages/salesforce-adapter/src/transformers/transformer.ts:1315]: Took me some time to figure this out, but I'm pretty sure we'll need to fix the way the salesforce-adapter e2e tests are defined in order for reference resolution to work correctly. From what I see, the issue is that the e2e test doesn't load the workspace - and the workspace is responsible for populating the `ReferenceExpression`'s `value`, which in turn is used to convert the expression back to a string before deploy. 
The relevant code in the E2E flow - `createElement` (for example in [this flow](https://github.com/salto-io/salto/blob/master/packages/salesforce-adapter/e2e_test/adapter.test.ts#L1581)) calls `adapter.deploy` directly ([here](https://github.com/salto-io/salto/blob/master/packages/salesforce-adapter/e2e_test/utils.ts#L154)), without running `loadWorkspace` or its equivalents.
The part that's responsible for populating the references before deploy is called as part of `validateWorkspace` - specifically `resolveReferenceExpression`, [here](https://github.com/salto-io/salto/blob/master/packages/workspace/src/expressions.ts#L102).

This code is not a good solution of course - only used it to verify this is the root cause (and it did eliminate all remaining test failures for the salesforce adapter E2E).
@ori-moisis @omrilit what's the best way to address this? The two main options I can think of are either to mock reference expression resolution for testing (far from ideal for e2e), or to replace `createElement` with a more standard deploy flow (probably has performance implications and may introduce other breakages). Are there other options I'm missing / is one of these preferred?

Comment (ori-moisis) [reply packages/salesforce-adapter/src/transformers/transformer.ts:1315]: This isn't the first time we encounter this in e2e: [here is another hacky fix for the same thing](https://github.com/salto-io/salto/blob/master/packages/salesforce-adapter/e2e_test/adapter.test.ts#L1566-L1578).

To actually "solve" this from a system arch standpoint the only way I think would be to force adapters to return resolved references, the benefit would be in symmetry, but it might put an unnecessary burden on adapters since we don't actually need the references to be resolved on fetch.
Personally I am in favor of doing this in some way because I think it makes a lot of sense that an adapter should expect to get the same elements it returned from fetch in calls to deploy.

Assuming however that we are not doing this right now, with the current state of things passing unresolved references to deploy can be considered invalid input (i.e - the test is testing something that should not be supported), resolving the references before deploying is the only real option.

from a theoretical standpoint I think it is ok for the e2e test to generate input to the adapter with some mocking, an e2e test in the adapter's scope really does start with resolved elements since that is this package's input.

Now to the part where I actually suggest something concrete to do - I think the best option forward right now would be similar to your first suggestion to "mock reference expression resolution", but not so much mock as actually resolve the references, replacing [the existing hack](https://github.com/salto-io/salto/blob/master/packages/salesforce-adapter/e2e_test/adapter.test.ts#L1566-L1578) such that we basically resolve all references on the fetch resolve before the add / update / remove tests. I do think we should mimic core's behavior better though (currently the hack fully replaces the reference expression with the api name, I think we should only fill the reference value like `core` would do and let the adapter `deploy` do the rest).
~Unfortunately, the reference resolution part happens in core, and at the moment core statically depends on the adapters so trying to import core from the adapter would cause a circular dependency.~
Actually, it seems like the resolve function was moved to the `workspace` package, so it can be used in the adapter without creating a dependency cycle.
So basically, at the end of the fetch test, we should create a resolved copy of the fetched elements and use that one for all the deploy tests

File: packages/cli/e2e_test/NACL/salto.config/salesforce.nacl [hunk 4]
```diff
@@ -5,6 +5,7 @@ salesforce {
     "ReportFolder",
     "Dashboard",
     "DashboardFolder",
+    "Profile",
```
Comment (ori-moisis) [inline packages/cli/e2e_test/NACL/salto.config/salesforce.nacl:8]: seems like we should remove everything from the skip list here.
if I understand the role of this file correctly, we would actually want to check that we are able to fetch all the types, even if they are skipped by default (we should still support the types that are skipped here)

Comment (netama) [reply packages/cli/e2e_test/NACL/salto.config/salesforce.nacl:8]: I think the way it's written today we would have to avoid the skiplist for other tests too in order to avoid breakages from unrelated diffs (like the one I fixed here). it seems there's an implicit assumption that the only difference between this and the default config is in `enableHideTypesInNacls`. (we can also try to make this a real diff operation / make the plan comparisons ignore additional differences)

File: packages/salesforce-adapter/src/transformers/transformer.ts (hunk 3 above)
Comment (netama) [reply packages/salesforce-adapter/src/transformers/transformer.ts]: looks like importing `workspace` still causes a dependency cycle because it relies on the adapters. I extended the existing workaround for now, lmk if that's good enough or if we want to do something else.

File: packages/cli/e2e_test/NACL/salto.config/salesforce.nacl (hunk 4 above)
Comment (ori-moisis) [reply packages/cli/e2e_test/NACL/salto.config/salesforce.nacl:8]: I didn't really understand why Profile was added to the skip list here (not sure what was fixed by it), this isn't really related to your change anyway so it is ok to merge like this I think.
I was just trying to raise a point about the e2e test maybe not doing the best job if it is ignoring some types that we are supposed to support, but if this isn't the case and the test is fine then feel free to ignore the comment

File: packages/salesforce-adapter/e2e_test/adapter.test.ts [hunk 5]
```diff
@@ -167,7 +176,7 @@ describe('Salesforce adapter E2E with real account', () => {
 
         // Test lookup reference_to annotation
         expect(
-          lead.fields.OwnerId.annotations[constants.FIELD_ANNOTATIONS.REFERENCE_TO]
+          extractReferenceTo(lead.fields.OwnerId.annotations)
```
Comment (ori-moisis) [inline packages/salesforce-adapter/e2e_test/adapter.test.ts:180]: shouldn't we expect this to be a reference? otherwise we are not testing your new feature...

File: packages/salesforce-adapter/e2e_test/adapter.test.ts [hunk 6]
```diff
@@ -1563,19 +1573,33 @@ describe('Salesforce adapter E2E with real account', () => {
...
-            const elem = findElement(result, ref.elemId)
-            return elem ? apiName(elem) : undefined
+          // Resolve reference expression before deploy
+          const normalizeReference = (
+            ref: ReferenceExpression | string | undefined
+          ): string | undefined => {
+            if (isReferenceExpression(ref)) {
+              const elem = findElement(result, ref.elemId)
+              return elem
+                // adding fallback for partially-resolved elements
```
Comment (ori-moisis) [inline packages/salesforce-adapter/e2e_test/adapter.test.ts:1583]: what does "partially-resolved elements" mean?
is this for split elements (custom objects)? if so, at the very least the comment should say so but probably better to actually support this case (get all the elements with the elem ID we are looking for and get the api name from the one that has it)

File: packages/salesforce-adapter/src/filters/reference_annotations.ts [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+      }
+    }
+    return ref
+  }
+
+  elements
+    .filter(isObjectType)
+    .flatMap((obj: ObjectType) => Object.values(obj.fields))
+    .filter((field: Field) => field.annotations[annotationName] !== undefined)
+    .forEach((field: Field): void => {
```
Comment (ori-moisis) [inline packages/salesforce-adapter/src/filters/reference_annotations.ts:54]:
```suggestion
    .flatMap(obj => Object.values(obj.fields))
    .filter(field => field.annotations[annotationName] !== undefined)
    .forEach(field => {
```

File: packages/salesforce-adapter/src/filters/reference_annotations.ts [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+)
+
+/**
+ * Convert referenceTo and foreignKeyDomain annotations into reference expressions.
+ */
+const filter: FilterCreator = () => ({
+  onFetch: async (elements: Element[]) => {
+    const typeToElemID = apiNameToElemID(elements)
+    convertAnnotationsToReferences(elements, typeToElemID, REFERENCE_TO)
+    convertAnnotationsToReferences(elements, typeToElemID, FOREIGN_KEY_DOMAIN)
```
Comment (ori-moisis) [inline packages/salesforce-adapter/src/filters/reference_annotations.ts:75]: nit: I know we don't really optimize for performance in these filters but running on all fields twice instead of handling both annotations just seems wasteful

File: packages/salesforce-adapter/e2e_test/adapter.test.ts (hunk 5 above)
Comment (netama) [reply packages/salesforce-adapter/e2e_test/adapter.test.ts:180]: we are still checking them, just using a helper function. 
if this is about the fallback when it's not a reference - originally the fallback in the function was to `undefined`, the reason I changed it was that two of the references used in the test (`Opportunity` and `Case`) are not returned from salesforce, so they are not converted to references (in the real code as well). I'll change the test to be more explicit about it and only avoid resolving in these two cases, but the flow is covered.
